### PR TITLE
ElasticSearch Query Editor - Scripted Metric

### PR DESCRIPTION
**Pull request by shanielh**
Further to #3441 

A pull request to add `Scripted Metric` option to query builder.

Couldn't do the "custom aggregation", too hard for me, Would love to get some guidance from someone on that one.

scripted metric

### CODE REVIEW

File: public/app/plugins/datasource/elasticsearch/partials/metricAgg.html [hunk 1]
```diff
@@ -62,5 +65,20 @@
 			</ul>
 			<div class="clearfix"></div>
 		</div>
+
+		<div ng-if="agg.type === 'scripted_metric'">
+			<div class="tight-form" ng-repeat="option in scriptedMetricOptions" ng-class="{last: $last}">
```
Comment (shanielh) [inline public/app/plugins/datasource/elasticsearch/partials/metricAgg.html]: Can I use here `aggDef`?

**Comment by torkelo** [comment]
this looks like an experimental Elasticsearch feature so it might have to wait before being included in Grafana

**Comment by shanielh** [comment]
This feature is available since elastic 1.4 (https://www.elastic.co/guide/en/elasticsearch/reference/1.4/search-aggregations-metrics-scripted-metric-aggregation.html). It's still considered experimental, but it's old feature. Maybe we can work on solution for "advanced" users only?

**Comment by Qtax** [comment]
Maybe we could also use the proposed ES datasource "enable scripts" setting in https://github.com/grafana/grafana/pull/3438 for this too? In which case we could hide this unless the option is enabled (disabled by default).

**Comment by Qtax** [comment]
Imho it could be nice to see a `lang` setting. (In case you don't want to change the default on ES.)
Don't know if ppl really need it, but I would like to see it somewhere in Grafana at least. If not desired on every metric, then possibly on the datasource.

**Comment by shanielh** [comment]
@Qtax :+1: I also need the scripting abilities because 2.6 disabled the manual query editor.

**Comment by shanielh** [comment]
@Qtax @torkelo Hi, I haven't added the language feature because I figured out from the other PR (#3438) that @torkelo doesn't want this setting.

But I did added the "Experimental Features" option in the ElasticSearch Datasource configuration panel : 

image

The "Scripted" Metric feature will be shown only when that option is checked.

I hope that after this edit, This pull request will be merged. @torkelo - If there is anything you want me to change, Please tell me.

**Comment by barvinograd** [comment]
+1

**Comment by shanielh** [comment]
@torkelo Any update regarding this? It will be helpful for me. 

I cannot updated to 2.6 because `Editor Mode` is not supported in that version and I'm looking forward to upgrade for other bugs fixes.

Elastic 2.2 is out and scripted metric is [still](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-scripted-metric-aggregation.html) supported.

**Comment by ozarvatz** [comment]
:+1:

**Comment by eliranmoyal** [comment]
+1

**Comment by OrenDor** [comment]
:+1:

**Comment by jisaitua** [comment]
+1

**Comment by torkelo** [comment]
closing this as this PR has drifted a bit too far from master (many conflicts). and needs a bit more reactions / love / demand on the feature request issue: https://github.com/grafana/grafana/issues/3441

**Pull request closed without merging**